cmd/jaeger/internal/extension/remotesampling/extension.go

```diff
@@ -0,0 +1,259 @@
+// Copyright (c) 2024 The Jaeger Authors.
+// SPDX-License-Identifier: Apache-2.0
+
+package remotesampling
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"net"
+	"net/http"
+	"sync"
+
+	"go.opentelemetry.io/collector/component"
+	"go.opentelemetry.io/collector/extension"
+	"go.uber.org/zap"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/health"
+	"google.golang.org/grpc/health/grpc_health_v1"
+
+	"github.com/jaegertracing/jaeger/cmd/collector/app/sampling"
+	"github.com/jaegertracing/jaeger/cmd/collector/app/sampling/strategystore"
+	"github.com/jaegertracing/jaeger/cmd/jaeger/internal/extension/jaegerstorage"
+	"github.com/jaegertracing/jaeger/pkg/clientcfg/clientcfghttp"
+	"github.com/jaegertracing/jaeger/pkg/metrics"
+	"github.com/jaegertracing/jaeger/plugin/sampling/leaderelection"
+	"github.com/jaegertracing/jaeger/plugin/sampling/strategystore/adaptive"
+	"github.com/jaegertracing/jaeger/plugin/sampling/strategystore/static"
+	"github.com/jaegertracing/jaeger/proto-gen/api_v2"
+	"github.com/jaegertracing/jaeger/storage"
+	"github.com/jaegertracing/jaeger/storage/samplingstore"
+)
+
+var _ extension.Extension = (*rsExtension)(nil)
+
+const defaultResourceName = "sampling_store_leader"
+
+type rsExtension struct {
+	cfg           *Config
+	telemetry     component.TelemetrySettings
+	httpServer    *http.Server
+	grpcServer    *grpc.Server
+	strategyStore strategystore.StrategyStore
+	shutdownWG    sync.WaitGroup
+}
+
+func newExtension(cfg *Config, telemetry component.TelemetrySettings) *rsExtension {
+	return &rsExtension{
+		cfg:       cfg,
+		telemetry: telemetry,
+	}
+}
+
+func GetExtensionConfig(host component.Host) (AdaptiveConfig, error) {
+	var comp component.Component
+	for id, ext := range host.GetExtensions() {
+		if id.Type() == componentType {
+			comp = ext
+			break
+		}
+	}
+	if comp == nil {
+		return AdaptiveConfig{}, fmt.Errorf(
+			"cannot find extension '%s' (make sure it's defined earlier in the config)",
+			componentType,
+		)
+	}
+	return comp.(*rsExtension).cfg.Adaptive, nil
+}
+
+// GetSamplingStorage retrieves a storage factory from `jaegerstorage` extension
+// and uses it to create a sampling store and a loader/follower implementation.
+func GetSamplingStorage(
+	samplingStorage string,
+	host component.Host,
+	opts adaptive.Options,
```

Comment:
I don't think we should be passing any arguments here aside from Host.

The `remotesampling` extension needs storage and leader/follower manager to function. It should instantiate those during its Start method. It will use storage name and options to do so. The processor does not need to know any of that, all it needs is to get those two components from the extension, the storage and leader/follower.

The reason it needs Host is because this is a static method, so we can use the Host to locate remotesampling extension and ask it to provide these two components, similar to how jaegerquery extension asks jaegerstorage for the storage (except in this case we don't even need the storage name):
```
func (s *server) Start(ctx context.Context, host component.Host) error {
	f, err := jaegerstorage.GetStorageFactory(s.config.TraceStoragePrimary, host)
```